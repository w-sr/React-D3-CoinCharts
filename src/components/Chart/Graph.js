import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import isEqual from "lodash.isequal";
import { area as d3Area, line as d3Line } from "d3-shape";
import { extent } from "d3-array";
import { easeCubicOut } from "d3-ease";
import { interpolatePath } from "d3-interpolate-path";
import { scaleLinear, scaleTime } from "d3-scale";
import { select } from "d3-selection";
import "d3-transition";

import { PROPTYPES } from "../../constants";

export const GRAPH_PADDING_BOTTOM = 15; // spacing between lowest datum & border
export const GRAPH_PADDING_TOP = 15; // spacing between highest datum & border
const INITIAL_STATE = {
  previousColor: undefined,
  previousScaledData: [],
  scaledData: [],
  skipTransition: false,
};
const TRANSITION = { duration: 500, ease: easeCubicOut };

const StyledGraph = styled.svg`
  height: 100%;
  width: 100%;

  .area {
    pointer-events: none;
    stroke: none;
  }

  .line {
    fill: none;
    pointer-events: none;
    stroke-width: 2;
  }
`;

class Graph extends Component {
  static scaleData(data, height, width) {
    const scalePriceToY = scaleLinear()
      .range([height - GRAPH_PADDING_BOTTOM, GRAPH_PADDING_TOP])
      .domain(extent(data, d => d.price));

    const scaleTimeToX = scaleTime()
      .range([0, width])
      .domain(extent(data, d => d.time));

    return data.map(({ price, time }) => ({
      price: scalePriceToY(price),
      time: scaleTimeToX(time),
    }));
  }

  constructor(props) {
    super(props);
    this.state = INITIAL_STATE;
  }

  componentWillReceiveProps(nextProps) {
    const { data: nextData, height: nextHeight, width: nextWidth } = nextProps;
    const { color, width } = this.props;

    // Don't update if next set of data is not ready
    if (nextData.length === 0) {
      return;
    }

    this.setState(prevState => {
      const { scaledData } = prevState;
      const nextScaledData = Graph.scaleData(nextData, nextHeight, nextWidth);
      const previousScaledData =
        scaledData.length > 0 ? scaledData : nextScaledData.map(({ time }) => ({ price: nextHeight, time }));

      return {
        skipTransition: width !== nextWidth,
        previousColor: color,
        previousScaledData,
        scaledData: nextScaledData,
      };
    });
  }

  shouldComponentUpdate(nextProps) {
    return !isEqual(nextProps, this.props);
  }

  componentDidUpdate() {
    const { color, height } = this.props;
    const { previousColor = color, previousScaledData, scaledData, skipTransition } = this.state;
    const graph = select(this.svgNode);
    const transitionDuration = skipTransition ? 0 : TRANSITION.duration;

    const area = d3Area()
      .x(d => d.time)
      .y0(height)
      .y1(d => d.price);
    const line = d3Line()
      .x(d => d.time)
      .y(d => d.price);

    const previousAreaGraph = area(previousScaledData);
    const previousLineGraph = line(previousScaledData);
    const areaGraph = area(scaledData);
    const lineGraph = line(scaledData);

    graph.selectAll("path").remove();

    graph
      .append("path")
      .attr("class", "area")
      .attr("d", previousAreaGraph)
      .style("fill", previousColor.fill)
      .transition()
      .duration(transitionDuration)
      .ease(TRANSITION.ease)
      .attrTween("d", interpolatePath.bind(null, previousAreaGraph, areaGraph))
      .style("fill", color.fill);

    graph
      .append("path")
      .attr("class", "line")
      .attr("d", previousLineGraph)
      .style("stroke", previousColor.stroke)
      .transition()
      .duration(transitionDuration)
      .ease(TRANSITION.ease)
      .attrTween("d", interpolatePath.bind(null, previousLineGraph, lineGraph))
      .style("stroke", color.stroke);
  }

  render() {
    const svgRef = node => {
      this.svgNode = node;
    };
    return (
      <StyledGraph data-testid="Graph">
        <g ref={svgRef} />
      </StyledGraph>
    );
  }
}

Graph.propTypes = {
  color: PROPTYPES.COLOR.isRequired,
  data: PROPTYPES.PRICE_DATA.isRequired,
  height: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
};

export default Graph;
