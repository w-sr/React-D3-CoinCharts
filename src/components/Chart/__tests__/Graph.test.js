import React from "react";
import { render } from "react-testing-library";

import Graph from "../Graph";

const TEST_PROPS = {
  color: {
    fill: "#333",
    stroke: "#666",
  },
  data: [
    { price: 10, time: new Date(1000) },
    { price: 20, time: new Date(2000) },
    { price: 30, time: new Date(3000) },
    { price: 40, time: new Date(4000) },
  ],
  height: 1,
  width: 1,
};

describe("<Graph />", () => {
  it("renders without crashing", () => {
    const { container } = render(<Graph {...TEST_PROPS} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders SVG SVG <path/>'s when `props.data` is not empty", () => {
    // Set <Graph/> to be initially empty & update props after initial render
    const { container } = render(<Graph {...TEST_PROPS} data={[]} />);
    render(<Graph {...TEST_PROPS} />, { container });

    expect(container.firstChild.querySelectorAll("path")).not.toHaveLength(0);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("does not render SVG <path/>'s when `props.data` is []", () => {
    const { container } = render(<Graph {...TEST_PROPS} data={[]} />);
    expect(container.firstChild.querySelectorAll("path")).toHaveLength(0);
    expect(container.firstChild).toMatchSnapshot();
  });
});
