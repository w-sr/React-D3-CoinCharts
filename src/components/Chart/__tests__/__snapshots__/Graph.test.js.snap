// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Graph /> does not render SVG <path/>'s when \`props.data\` is [] 1`] = `
<svg
  class="sc-bdVaJa frmpus"
  data-testid="Graph"
>
  <g />
</svg>
`;

exports[`<Graph /> renders SVG SVG <path/>'s when \`props.data\` is not empty 1`] = `
<svg
  class="sc-bdVaJa frmpus"
  data-testid="Graph"
>
  <g>
    <path
      class="area"
      d="M0,1L0.3333333333333333,1L0.6666666666666666,1L1,1L1,1L0.6666666666666666,1L0.3333333333333333,1L0,1Z"
      style="fill: #333;"
    />
    <path
      class="line"
      d="M0,1L0.3333333333333333,1L0.6666666666666666,1L1,1"
      style="stroke: #666;"
    />
  </g>
</svg>
`;

exports[`<Graph /> renders without crashing 1`] = `
<svg
  class="sc-bdVaJa frmpus"
  data-testid="Graph"
>
  <g />
</svg>
`;
